### DISCUSSION TITLE
Save schema as PNG, connection tags and grid display performance

### DISCUSSION THREAD

**Discussion by yegorf** [eckig/graph-editor#200] [Q&A] [answered]
Hello, Steffen.

I have 3 questions:
1. Is there any "native" solution to save the whole schema as an image? I have implemented the custom solution - move the Container's viewport to a specific page and make its snapshot. But I don't really like this solution.
![image](https://github.com/eckig/graph-editor/assets/29003239/bb9e5f55-2174-49b8-96f8-3d6606537bdd)

2. I have some Tag nodes and I want to connect them to existing connections. Something like this: 
![Tags](https://github.com/eckig/graph-editor/assets/29003239/65f5f151-cffa-4ca9-88fb-50db650d4eb0)

3. I noticed that when working with large editor area sizes, some UI friezes occur, but only if the grid is displayed. Without a grid, everything works smoothly. Is this known and expected behavior? Can something be done to improve UI performance on grid render?
![image](https://github.com/eckig/graph-editor/assets/29003239/b32bbba0-bed9-4794-9407-2c070bec0920)

Thank you.

**Comment by eckig** [comment, marked as answer]
This is how I make screenshots of the entire graph:
```
        GraphEditor editor = ...;
        GraphEditorView editorView = editor.getView();
        SnapshotParameters params = new SnapshotParameters();
        Rectangle2D viewport = GraphUtils.getBounds(model);
        viewport = new Rectangle2D(viewport.getMinX() + editorView.getLayoutX(),
                viewport.getMinY() + editorView.getLayoutY(), viewport.getWidth(), viewport.getHeight());

        params.setViewport(viewport);

        editorView.snapshot(params, null);
```

For the second question: You could add a small `GNode` inbetween where both connections can come together.

For the third question: No, this is new. Maybe I can investigate this..

**Reply by yegorf** [reply]
Thanks a lot for your answer. 

1) Your snapshot logic works perfect. I will modify my each page snapshot logic using your approach. Using GraphEditorView instead of GraphEditorContainer makes it much better.

2) I have tried something like this, but I don't really understand how to implement it. As I understand the workflow: 1. User adds a Tag node with 4 connectors on each side. 2. The user drags the connection from the Tag node to the existing connection line. 3. In the moment when the cursor touches the line, we create a small node with 3 connectors and connect all 3 lines to it.
Am I right with this flow? If yes, could you please give some suggestions about step (3)?

3) I will let you know if I find out more about this problem.

**Reply by eckig** [reply]
Basically you said everything to do step 2: If you use a CompoundCommand to remove the previous connections and register the new ones, it should work.

**Reply by yegorf** [reply]
Got it, thank you very much.
